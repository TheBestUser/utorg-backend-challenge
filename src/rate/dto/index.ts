export * from './post-rate.dto';
