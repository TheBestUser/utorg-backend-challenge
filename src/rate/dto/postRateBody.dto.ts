export class PostRateBodyDto {
  readonly from: string;
  readonly to: string;
  readonly fromAmount?: string;
  readonly toAmount?: string;
}
