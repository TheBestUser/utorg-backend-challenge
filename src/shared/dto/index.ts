export * from './trade-rate.dto';
