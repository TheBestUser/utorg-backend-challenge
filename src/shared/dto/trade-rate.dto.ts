import { IsNumberString } from 'class-validator';
import { Transform } from 'class-transformer';
import { ceilEthers, floorEthers } from '../utils';
import { TradeRate } from '../interfaces';

export class TradeRateDto implements TradeRate {
  @IsNumberString()
  @Transform(({ value }) => ceilEthers(value, 6))
  fromAmount: string;

  @IsNumberString()
  @Transform(({ value }) => floorEthers(value, 6))
  toAmount: string;
}
