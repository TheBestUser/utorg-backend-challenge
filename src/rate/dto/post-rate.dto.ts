import { IsEthereumAddress, IsNumberString, IsOptional } from 'class-validator';
import { Transform } from 'class-transformer';
import { floorEthers } from '../../shared';
import { DEFAULT_PRECISION } from '../rate.constant';

export class PostRateDto {
  @IsEthereumAddress()
  from: string;

  @IsEthereumAddress()
  to: string;

  @Transform(({ value }) => floorEthers(value, DEFAULT_PRECISION))
  @IsOptional()
  @IsNumberString()
  fromAmount?: string;

  @Transform(({ value }) => floorEthers(value, DEFAULT_PRECISION))
  @IsOptional()
  @IsNumberString()
  toAmount?: string;
}
