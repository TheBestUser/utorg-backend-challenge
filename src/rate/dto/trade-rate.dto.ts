import { IsNumberString } from 'class-validator';
import { Transform, Type } from 'class-transformer';
import { ceilEthers, floorEthers } from '../../shared';
import { DEFAULT_PRECISION } from '../rate.constant';

export class TradeRateDto {
  @Type(() => String)
  @Transform(({ value }) => ceilEthers(value, DEFAULT_PRECISION))
  @IsNumberString()
  fromAmount: string;

  @Type(() => String)
  @Transform(({ value }) => floorEthers(value, DEFAULT_PRECISION))
  @IsNumberString()
  toAmount: string;
}
