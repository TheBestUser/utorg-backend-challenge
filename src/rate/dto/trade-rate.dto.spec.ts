import { plainToInstance } from 'class-transformer';
import { TradeRateDto } from './trade-rate.dto';

describe('TradeRateDto', () => {
  test.each`
    name                   | fromAmount
    ${'should round up'}   | ${'1000.1234567'}
    ${'should round up'}   | ${'999.99999999'}
    ${'should round down'} | ${'0'}
    ${'should round down'} | ${'0.0'}
    ${'should round down'} | ${'1000.000000'}
    ${'should round down'} | ${'1000.1234560'}
  `('$name: $fromAmount', ({ fromAmount }) => {
    expect(
      plainToInstance(TradeRateDto, {
        fromAmount,
      }),
    ).toMatchSnapshot('fromAmount');
  });

  test.each(['0', '0.0', '1000.000000', '1000.1234567', '999.99999999'])(
    'should round down: %s',
    (toAmount) => {
      expect(
        plainToInstance(TradeRateDto, {
          toAmount,
        }),
      ).toMatchSnapshot('toAmount');
    },
  );

  it('real example', () => {
    expect(
      plainToInstance(TradeRateDto, {
        fromAmount: '1000.123456',
        toAmount: '3.66425283767606108',
      }),
    ).toMatchInlineSnapshot(`
      TradeRateDto {
        "fromAmount": "1000.123456",
        "toAmount": "3.664252",
      }
    `);
  });
});
