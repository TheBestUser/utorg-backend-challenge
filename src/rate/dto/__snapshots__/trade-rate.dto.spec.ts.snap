// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TradeRateDto should round down: 0.0: fromAmount 1`] = `
TradeRateDto {
  "fromAmount": "0.000000",
}
`;

exports[`TradeRateDto should round down: 0.0: toAmount 1`] = `
TradeRateDto {
  "toAmount": "0.000000",
}
`;

exports[`TradeRateDto should round down: 0: fromAmount 1`] = `
TradeRateDto {
  "fromAmount": "0.000000",
}
`;

exports[`TradeRateDto should round down: 0: toAmount 1`] = `
TradeRateDto {
  "toAmount": "0.000000",
}
`;

exports[`TradeRateDto should round down: 999.99999999: toAmount 1`] = `
TradeRateDto {
  "toAmount": "999.999999",
}
`;

exports[`TradeRateDto should round down: 1000.000000: fromAmount 1`] = `
TradeRateDto {
  "fromAmount": "1000.000000",
}
`;

exports[`TradeRateDto should round down: 1000.000000: toAmount 1`] = `
TradeRateDto {
  "toAmount": "1000.000000",
}
`;

exports[`TradeRateDto should round down: 1000.1234560: fromAmount 1`] = `
TradeRateDto {
  "fromAmount": "1000.123456",
}
`;

exports[`TradeRateDto should round down: 1000.1234567: toAmount 1`] = `
TradeRateDto {
  "toAmount": "1000.123456",
}
`;

exports[`TradeRateDto should round up: 999.99999999: fromAmount 1`] = `
TradeRateDto {
  "fromAmount": "1000.000000",
}
`;

exports[`TradeRateDto should round up: 1000.1234567: fromAmount 1`] = `
TradeRateDto {
  "fromAmount": "1000.123457",
}
`;
